File: app/src/utils/deposit_collateral.ts
import * as anchor from '@coral-xyz/anchor';
import { Connection, PublicKey, SystemProgram } from '@solana/web3.js';
import { TOKEN_PROGRAM_ID, ASSOCIATED_TOKEN_PROGRAM_ID } from '@solana/spl-token';
import fallIdl from '../idl/fall.json';
import BN from 'bn.js';
import { 
  AUTHORITY_SEED,
  LENDING_SEED,
  LENDING_AUTHORITY_SEED,
  BORROW_TOKEN_SEED,
  COLLATERAL_TOKEN_SEED,
  BORROWER_BORROW_BLOCK_HEIGHT_TOKEN_SEED,
  BORROWER_AUTHORITY_SEED,
} from '../utils/constants';

export interface DepositCollateralResult {
  tx: string;
  accounts: {
    poolPda: PublicKey;
    poolAuthority: PublicKey;
    lendingPool: PublicKey;
    lendingPoolAuthority: PublicKey;
    borrowReceiptTokenMint: PublicKey;
    collateralReceiptTokenMint: PublicKey;
    borrowerBorrowBlockHeightTokenMint: PublicKey;
    borrowerBorrowReceiptToken: PublicKey;
    borrowerCollateralReceiptToken: PublicKey;
    borrowerBorrowBlockHeightReceiptToken: PublicKey;
  };
}

export async function depositCollateral(
  wallet: any,
  connection: Connection,
  poolPda: PublicKey,
  mintA: PublicKey,
  mintB: PublicKey,
  collateralAmount: BN,
): Promise<DepositCollateralResult> {
  try {
    console.log('Executing borrow...');
    const provider = new anchor.AnchorProvider(
      connection,
      wallet,
      { preflightCommitment: "confirmed" }
    );

    const program = new anchor.Program(
      fallIdl ,
      provider
    );

    // 获取 Pool 信息
    const pool = await program.account.pool.fetch(poolPda);
    console.log('Pool info:', {
      amm: pool.amm.toString(),
      mintA: pool.mintA.toString(),
      mintB: pool.mintB.toString(),
    });

    // 获取 Pool Authority PDA
    const [poolAuthority] = PublicKey.findProgramAddressSync(
      [
        pool.amm.toBuffer(),
        mintA.toBuffer(),
        mintB.toBuffer(),
        Buffer.from(AUTHORITY_SEED)
      ],
      program.programId
    );

    // 获取 Lending Pool PDA
    const [lendingPool] = PublicKey.findProgramAddressSync(
      [
        poolPda.toBuffer(),
        mintA.toBuffer(),
        mintB.toBuffer(),
        Buffer.from(LENDING_SEED)
      ],
      program.programId
    );

    // 获取 Lending Pool Authority PDA
    const [lendingPoolAuthority] = PublicKey.findProgramAddressSync(
      [
        poolPda.toBuffer(),
        mintA.toBuffer(),
        mintB.toBuffer(),
        Buffer.from(LENDING_AUTHORITY_SEED)
      ],
      program.programId
    );

    // 获取 Receipt Token Mints
    const [borrowReceiptTokenMint] = PublicKey.findProgramAddressSync(
      [
        poolPda.toBuffer(),
        mintA.toBuffer(),
        Buffer.from(BORROW_TOKEN_SEED)
      ],
      program.programId
    );

    const [collateralReceiptTokenMint] = PublicKey.findProgramAddressSync(
      [
        poolPda.toBuffer(),
        mintB.toBuffer(),
        Buffer.from(COLLATERAL_TOKEN_SEED)
      ],
      program.programId
    );

    // 获取 Borrower Borrow Block Height Token Mint
    const [borrowerBorrowBlockHeightTokenMint] = PublicKey.findProgramAddressSync(
      [
        poolPda.toBuffer(),
        Buffer.from(BORROWER_BORROW_BLOCK_HEIGHT_TOKEN_SEED)
      ],
      program.programId
    );

    // 获取相关 Token Accounts
    const poolAccountA = await anchor.utils.token.associatedAddress({
      mint: mintA,
      owner: poolAuthority
    });

    const poolAccountB = await anchor.utils.token.associatedAddress({
      mint: mintB,
      owner: poolAuthority
    });

    const lendingPoolTokenA = await anchor.utils.token.associatedAddress({
      mint: mintA,
      owner: lendingPoolAuthority
    });

    const lendingPoolTokenB = await anchor.utils.token.associatedAddress({
      mint: mintB,
      owner: lendingPoolAuthority
    });

    const borrowerTokenA = await anchor.utils.token.associatedAddress({
      mint: mintA,
      owner: provider.wallet.publicKey
    });

    const borrowerTokenB = await anchor.utils.token.associatedAddress({
      mint: mintB,
      owner: provider.wallet.publicKey
    });

    // 获取 Borrower Authority PDA
    const [borrowerAuthority] = PublicKey.findProgramAddressSync(
      [
        poolPda.toBuffer(),
        provider.wallet.publicKey.toBuffer(),
        Buffer.from(BORROWER_AUTHORITY_SEED)
      ],
      program.programId
    );
    console.log('Borrower Authority:', borrowerAuthority.toString());

    const borrowerBorrowReceiptToken = await anchor.utils.token.associatedAddress({
      mint: borrowReceiptTokenMint,
      owner: borrowerAuthority
    });

    const borrowerCollateralReceiptToken = await anchor.utils.token.associatedAddress({
      mint: collateralReceiptTokenMint,
      owner: borrowerAuthority
    });

    const borrowerBorrowBlockHeightReceiptToken = await anchor.utils.token.associatedAddress({
      mint: borrowerBorrowBlockHeightTokenMint,
      owner: borrowerAuthority
    });

    // const modifyComputeUnits = anchor.web3.ComputeBudgetProgram.setComputeUnitLimit({ 
    //   units: 500_000
    // });

    console.log('Sending borrow transaction...');
    const modifyComputeUnits = anchor.web3.ComputeBudgetProgram.setComputeUnitLimit({ 
      units: 1_000_000  // 增加到 1M 单位
    });
    const tx = await program.methods
      .depositCollateral( collateralAmount)
      .accounts({
        pool: poolPda,
        poolAuthority,
        mintA,
        mintB,
        poolAccountA,
        poolAccountB,
        lendingPool,
        lendingPoolAuthority,
        lendingPoolTokenA,
        lendingPoolTokenB,
        borrowReceiptTokenMint,
        collateralReceiptTokenMint,
        borrowerBorrowBlockHeightTokenMint,
        borrower: provider.wallet.publicKey,
        borrowerTokenA,
        borrowerTokenB,
        borrowerBorrowReceiptToken,
        borrowerCollateralReceiptToken,
        borrowerBorrowBlockHeightReceiptToken,
        payer: provider.wallet.publicKey,
        tokenProgram: TOKEN_PROGRAM_ID,
        associatedTokenProgram: ASSOCIATED_TOKEN_PROGRAM_ID,
        systemProgram: SystemProgram.programId,
      }).preInstructions([modifyComputeUnits]).rpc();

    return {
      tx,
      accounts: {
        poolPda,
        poolAuthority,
        lendingPool,
        lendingPoolAuthority,
        borrowReceiptTokenMint,
        collateralReceiptTokenMint,
        borrowerBorrowBlockHeightTokenMint,
        borrowerBorrowReceiptToken,
        borrowerCollateralReceiptToken,
        borrowerBorrowBlockHeightReceiptToken,
      }
    };
  } catch (error) {
    console.error('Error in borrow:', error);
    if (error instanceof Error) {
      console.error('Error details:', {
        message: error.message,
        stack: error.stack
      });
    }
    throw error;
  }
}

File: fall/programs/fall/src/instructions/deposit_collateral.rs
use anchor_lang::prelude::*;
use anchor_spl::{
    token::{self, Mint, Token, TokenAccount, Transfer},
    associated_token::AssociatedToken,
};
use crate::constants::{ LENDING_AUTHORITY_SEED, BORROWER_AUTHORITY_SEED, COLLATERAL_TOKEN_SEED,BORROWER_BORROW_BLOCK_HEIGHT_TOKEN_SEED};
use crate::state::Pool;
use crate::instructions::utils::mint_and_freeze_token;

#[derive(Accounts)]
pub struct DepositCollateral<'info> {
    #[account(
        seeds = [
            pool.amm.as_ref(),
            pool.mint_a.key().as_ref(),
            pool.mint_b.key().as_ref(),
        ],
        bump,
    )]
    pub pool: Box<Account<'info, Pool>>,
    
    /// CHECK: Read only authority
    #[account(
        seeds = [
            pool.key().as_ref(),
            LENDING_AUTHORITY_SEED,
        ],
        bump,
    )]
    pub lending_pool_authority: AccountInfo<'info>,

    #[account(
        mut,
        associated_token::mint = pool.mint_b,
        associated_token::authority = lending_pool_authority,
    )]
    pub lending_pool_token_b: Box<Account<'info, TokenAccount>>,

    #[account(
        mut,
        seeds = [
            pool.key().as_ref(),
            COLLATERAL_TOKEN_SEED,
        ],
        bump,
    )]
    pub collateral_receipt_token_mint: Box<Account<'info, Mint>>,

    #[account(
        mut,
        seeds = [
            pool.key().as_ref(),
            BORROWER_BORROW_BLOCK_HEIGHT_TOKEN_SEED,
        ],
        bump,
    )]
    pub borrower_borrow_block_height_mint: Box<Account<'info, Mint>>,

    pub borrower: Signer<'info>,

    #[account(
        mut,
        associated_token::mint = pool.mint_b,
        associated_token::authority = borrower,
    )]
    pub borrower_token_b: Box<Account<'info, TokenAccount>>,

    /// CHECK: Read only authority
    #[account(
        seeds = [
            pool.key().as_ref(),
            borrower.key().as_ref(),
            BORROWER_AUTHORITY_SEED,
        ],
        bump,
    )]
    pub borrower_authority: AccountInfo<'info>,

    #[account(
        init_if_needed,
        payer = payer,
        associated_token::mint = collateral_receipt_token_mint,
        associated_token::authority = borrower_authority,
    )]
    pub borrower_collateral_receipt_token: Box<Account<'info, TokenAccount>>,

    #[account(
        init_if_needed,
        payer = payer,
        associated_token::mint = borrower_borrow_block_height_mint,
        associated_token::authority = borrower_authority,
    )]
    pub borrower_borrow_block_height_receipt_token: Box<Account<'info, TokenAccount>>,
    

    #[account(mut)]
    pub payer: Signer<'info>,

    pub token_program: Program<'info, Token>,
    pub associated_token_program: Program<'info, AssociatedToken>,
    pub system_program: Program<'info, System>,
}

pub fn deposit_collateral(
    ctx: Context<DepositCollateral>,
    collateral_amount: u64,  // 抵押的 token B 数量
) -> Result<()> {
    // 转移抵押物（token B）到借贷池 铸造抵押物 collateral_receipt_token
    token::transfer(
        CpiContext::new(
            ctx.accounts.token_program.to_account_info(),
            Transfer {
                from: ctx.accounts.borrower_token_b.to_account_info(),
                to: ctx.accounts.lending_pool_token_b.to_account_info(),
                authority: ctx.accounts.borrower.to_account_info(),
            },
        ),
        collateral_amount,
    )?;
    let authority_seeds = &[
        &ctx.accounts.pool.key().to_bytes(),
        LENDING_AUTHORITY_SEED,
        &[ctx.bumps.lending_pool_authority],
    ];
    let signer_seeds = &[&authority_seeds[..]];
    mint_and_freeze_token(
        &ctx.accounts.token_program,
        &ctx.accounts.collateral_receipt_token_mint,
        &ctx.accounts.borrower_collateral_receipt_token,
        &ctx.accounts.lending_pool_authority,
        signer_seeds,
        collateral_amount,
    )?;
    
    Ok(())
}
